src/clojure_sandbox/4clojure.clj

```diff
@@ -0,0 +1,605 @@
+;;     the type of collection it was given. You won't be allowed to inspect
+;;     their class or use the built-in predicates like list? - the point is to
+;;     poke at them and understand their behavior.
+;;
+;; Special Restrictions:
+;; class,type,Class,vector?,sequential?,list?,seq?,map?,set?,instance?,getClass
+(defn type-coll [coll]
+  (let [c (+ 3 (count coll))
+        coll (conj coll [:a :b] [:a :b] [:c :d])
+        nc (count coll)]
+    (cond
+      (= :b (get coll :a)) :map
+      (not= c nc) :set
+      (= [:c :d] (first coll)) :list
+      (= [:c :d] (last coll)) :vector)))
+
+(comment (type-coll {:a 1, :b 2}))
+(comment (type-coll (range (rand-int 20))))
+(comment (type-coll [1 2 3 4 5 6]))
+(comment (type-coll #{10 (rand-int 5)}))
+(comment (map type-coll [{} #{} [] ()]))
+
+;; 66. Given two integers, write a function which returns the greatest common
+;;     divisor.
+;;
+;; https://en.wikipedia.org/wiki/Euclidean_algorithm
+(defn gcd [x y]
+  (loop [a (max x y)
+         b (min x y)]
+    (if (zero? b)
+      a
+      (recur b (mod a b)))))
+
+(comment (gcd 2 4))
+(comment (gcd 10 5))
+(comment (gcd 5 7))
+(comment (gcd 1023 858))
+
+;; 67. Write a function which returns the first x number of prime numbers.
+(defn prime-numbers [x]
+  (letfn [(prime? [n]
+            (cond
+              (<= n 1) false
+              (<= n 3) true
+              :else (loop [[i & remaining] (range 2 (inc (Math/sqrt n)))]
+                      (cond
+                        (nil? i) true
+                        (zero? (rem n i)) false
+                        :else (recur remaining)))))]
+    (take x (filter prime? (range)))))
+
+(comment (prime-numbers 2))
+(comment (prime-numbers 5))
+(comment (last (prime-numbers 100)))
+
+;; 68 on 4Clojure
+
+;; 69. Write a function which takes a function f and a variable number of maps.
+;;     Your function should return a map that consists of the rest of the maps
+;;     conj-ed onto the first. If a key occurs in more than one map, the
+;;     mapping(s) from the latter (left-to-right) should be combined with the
+;;     mapping in the result by calling (f val-in-result val-in-latter)
+(defn merge-with* [f map & maps]
+  (reduce
+   (fn [result map]
+     (reduce
+      (fn [result [key val]]
+        (if (contains? result key)
+          (update result key f val)
+          (assoc result key val)))
+      result
+      map))
+   map
+   maps))
+
+(comment (merge-with* * {:a 2, :b 3, :c 4} {:a 2} {:b 2} {:c 5}))
+(comment (merge-with* - {1 10, 2 20} {1 3, 2 10, 3 15}))
+(comment (merge-with* concat {:a [3], :b [6]} {:a [4 5], :c [8 9]} {:b [7]}))
+
+;; 70. Write a function which splits a sentence up into a sorted list of words.
+;;     Capitalization should not affect sort order and punctuation should be
+;;     ignored.
+(defn split-sort-words [s]
+  (sort
+   #(compare (str/lower-case %1) (str/lower-case %2))
+   (re-seq #"\w+" s)))
```

Comment:
comment: this `re-seq` usage is nice, I would have gone by just split to convert sentence into words. Creating a lazy seq here looks useful.
But, sort will convert any seq to an array, so do lazy sequences add any value here?

Reply:
I started with the split but then got stuck on manually dropping punctuations. When I looked at other people's solutions, `re-seq` seemed like a good approach. Besides, `re-seq` returns a lazy sequence.

> But, sort will convert any seq to an array, so do lazy sequences add any value here?

I don't think they will. `sort` internally creates a new Java array from the given sequence and then uses Java's `Arrays#sort`, which should be eager..?